docs/developer/best-practices/stability.asciidoc

```diff
@@ -43,11 +43,30 @@ dependency list!
 [discrete]
 === Test coverage
 
-* Does the feature have sufficient unit test coverage? (does it handle
-storeinSessions?)
-* Does the feature have sufficient Functional UI test coverage?
-* Does the feature have sufficient Rest API coverage test coverage?
-* Does the feature have sufficient Integration test coverage?
+Testing UI code is hard. Here are our heuristics for maintaining Kibana's quality with tests.
+
+**Priority 1:** Our first priority is to minimize the total number of poor user experiences caused by bugs, and to prevent bugs that can cause harm to the user. "Harm" includes direct problems like data loss and data entering a bad state, as well as indirect problems like making a poor business decision based on misinformation presented by the UI.
+
+**Priority 2:** Our second priority is to define all correct UX behavior and guard against any UX regressions.
+
+**We are constrained by time.** We don’t have enough time to test all possible input and output permutations of the UX. Just think of the different ways a form can be interacted with and respond to input. We must spend our time prudently to address the above priorities.
+
+With each feature we ship, we have a **hard requirement** for test coverage of common library code, popular UIs, critical code (e.g. payload de/serialization, server-side APIs), and critical features (e.g. visualization, security). We have a **soft requirement** for test coverage of everything else.
+
+**We can make an exception** for the soft requirement if there’s a business case for shipping a feature on time. This requires buy-in from a lead and a PM, and a written explanation in the PR to support forensics. Any time we make an exception we immediately follow up with a separate PR to add the missing test coverage. 
```

Comment:
I'm really hesitant to suggest that we can make exceptions for testing in order to hit a deadline. This goes against [Kibana's Developer Principles](https://github.com/elastic/engineering/blob/master/kibana_dev_principles.md#automate-tests-through-ci), which states:

> It's imperative for the future success and stability of the project as well as for team velocity that all changes have extensive automated testing that give us confidence that ongoing churn on the codebase does not break existing functionality. We must achieve this level of testing without increasing the manual QA burden.

----

> It's natural for a developer to want to get an important change merged into the product so it can be released, and it's good for our users to have access to frequent improvements in our releases, but we must not sacrifice the quality of our changes in order to meet these pressures.

IMO, insufficient test coverage is sacrificing the quality of our changes, and yielding to these pressures can set a bad precedent for future features.

Reply:
Thanks for pointing this out, Larry. I think you're right.